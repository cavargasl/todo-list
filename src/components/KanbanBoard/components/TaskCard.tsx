import { type FunctionsTask, type Task } from "@/types"

import Button from "@/components/ui/Button"
import CheckBox from "@/components/ui/CheckBox"
import ArrowLeft from "@/components/icons/ArrowLeft"
import ArrowRight from "@/components/icons/ArrowRight"
import TrashIcon from "@/components/icons/TrashIcon"

type TaskCardProps = {
  task: Task
  functionsTask: Omit<FunctionsTask, "onAddTask">
  lastStage: boolean
}
export default function TaskCard({
  task,
  functionsTask,
  lastStage,
}: TaskCardProps) {
  function moveLeft() {
    if (task.stage === 0) return
    functionsTask.onMoveTask({ taskId: task.id, newStage: task.stage - 1 })
  }
  function moveRight() {
    if (lastStage) return
    functionsTask.onMoveTask({ taskId: task.id, newStage: task.stage + 1 })
  }
  return (
    <li
      aria-label="task"
      className="flex w-full gap-2 rounded-md bg-background p-2"
    >
      <CheckBox
        className="peer"
        defaultChecked={task.completed}
        onChange={() =>
          functionsTask.onEditTask({
            taskId: task.id,
            newTask: { ...task, completed: !task.completed },
          })
        }
      />
      <p className="w-full peer-checked:line-through">{task.name}</p>
      <div className="flex gap-1">
        <Button
          variant={"ghost"}
          size={"icon"}
          className="size-[20px] p-0"
          aria-label="move to left"
          onClick={moveLeft}
          disabled={task.stage === 0}
        >
          <ArrowLeft />
        </Button>
        <Button
          variant={"ghost"}
          size={"icon"}
          className="size-[20px] p-0"
          aria-label="move to right"
          onClick={moveRight}
          disabled={lastStage}
        >
          <ArrowRight />
        </Button>
        <Button
          variant={"ghost"}
          size={"icon"}
          className="size-[20px] p-0"
          aria-label="delete task"
          onClick={() => functionsTask.onDeleteTask(task.id)}
        >
          <TrashIcon />
        </Button>
      </div>
    </li>
  )
}
