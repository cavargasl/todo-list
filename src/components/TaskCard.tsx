import { Task } from "@/types";
import ArrowLeft from "./icons/ArrowLeft";
import ArrowRight from "./icons/ArrowRight";
import TrashIcon from "./icons/TrashIcon";
import Button from "./ui/Button";
import CheckBox from "./ui/CheckBox";

export default function TaskCard({ Task: { name, completed } }: { Task: Task }) {
  return (
    <article className="bg-background w-full p-2 rounded-md flex gap-2">
      <CheckBox className="peer" defaultChecked={completed} />
      <p className="w-full peer-checked:line-through">{name}</p>
      <div className="flex gap-1">
        <Button
          variant={"ghost"}
          size={"icon"}
          className="h-[20px] w-[20px] p-0"
        >
          <ArrowLeft />
        </Button>
        <Button
          variant={"ghost"}
          size={"icon"}
          className="h-[20px] w-[20px] p-0"
        >
          <ArrowRight />
        </Button>
        <Button
          variant={"ghost"}
          size={"icon"}
          className="h-[20px] w-[20px] p-0"
        >
          <TrashIcon />
        </Button>
      </div>
    </article>
  );
}
