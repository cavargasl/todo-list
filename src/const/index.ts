export const defaultColumns = ["To do", "In progress", "Done"];