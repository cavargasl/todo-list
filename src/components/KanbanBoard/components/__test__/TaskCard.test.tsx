import { defaultTasks } from "@/const"
import { cleanup, render, screen } from "@testing-library/react"
import userEvent from "@testing-library/user-event"

import TaskCard from "../TaskCard"

const defaultTask = defaultTasks[0]
const defaultTaskCompleted = defaultTasks[1]
const functionsTaskMock = {
  onDeleteTask: vi.fn(),
  onEditTask: vi.fn(),
  onMoveTask: vi.fn(),
}

describe("TaskCard", () => {
  beforeEach(() => {
    render(
      <TaskCard
        task={defaultTask}
        functionsTask={functionsTaskMock}
        lastStage={false}
      />
    )
  })

  test("renders the component ", () => {
    expect(screen.getByRole("listitem", { name: "task" })).toMatchSnapshot()
  })

  test("renders task name and checkbox", () => {
    const taskNameElement = screen.getByRole("paragraph")
    expect(taskNameElement).toBeInTheDocument()
    expect(taskNameElement).toHaveTextContent(defaultTask.name)
    expect(screen.getByRole("checkbox")).toBeInTheDocument()
  })

  test("renders component with completed style", () => {
    cleanup()
    render(
      <TaskCard
        task={defaultTaskCompleted}
        functionsTask={functionsTaskMock}
        lastStage
      />
    )
    expect(screen.getByRole("listitem", { name: "task" })).toMatchSnapshot()
    expect(screen.getByRole("paragraph")).toHaveStyle(
      "text-decoration-line: line-through"
    )
    expect(screen.getByRole("checkbox")).toBeChecked()
  })

  test("renders component with not completed style", () => {
    expect(screen.getByRole("paragraph")).not.toHaveStyle(
      "text-decoration-line: line-through"
    )
    expect(screen.getByRole("checkbox")).not.toBeChecked()
  })

  test("renders the buttons", () => {
    const buttons = screen.getAllByRole("button")
    expect(buttons).toHaveLength(3)
    expect(buttons[0]).toHaveAttribute("aria-label", "move to left")
    expect(buttons[1]).toHaveAttribute("aria-label", "move to right")
    expect(buttons[2]).toHaveAttribute("aria-label", "delete task")
  })

  describe("when clicking the buttons", () => {
    test("calls the function to move to left the task", async () => {
      const moveTaskLeftButton = screen.getByRole("button", {
        name: "move to left",
      })
      expect(moveTaskLeftButton).toBeInTheDocument()
      await userEvent.click(moveTaskLeftButton)
      expect(functionsTaskMock.onMoveTask).toHaveBeenCalled()
      expect(functionsTaskMock.onMoveTask).toHaveBeenCalledWith({
        taskId: defaultTask.id,
        newStage: defaultTask.stage - 1,
      })
    })

    test("disables the button to move to left when the task is in the first stage", async () => {
      cleanup()
      render(
        <TaskCard
          task={{ ...defaultTask, stage: 0 }}
          functionsTask={functionsTaskMock}
          lastStage={false}
        />
      )
      const moveTaskLeftButton = screen.getByRole("button", {
        name: "move to left",
      })
      expect(moveTaskLeftButton).toBeDisabled()
      let clickedFailed = false
      try {
        await userEvent.click(moveTaskLeftButton)
      } catch (error) {
        clickedFailed = true
      }
      expect(clickedFailed).toBe(true)
    })

    test("calls the function to move to right the task", async () => {
      const moveTaskRightButton = screen.getByRole("button", {
        name: "move to right",
      })
      expect(moveTaskRightButton).toBeInTheDocument()
      await userEvent.click(moveTaskRightButton)
      expect(functionsTaskMock.onMoveTask).toHaveBeenCalled()
      expect(functionsTaskMock.onMoveTask).toHaveBeenCalledWith({
        taskId: defaultTask.id,
        newStage: defaultTask.stage + 1,
      })
    })

    test("disables the button to move to right when the task is in the last stage", async () => {
      cleanup()
      render(
        <TaskCard
          task={defaultTask}
          functionsTask={functionsTaskMock}
          lastStage
        />
      )
      const moveTaskRightButton = screen.getByRole("button", {
        name: "move to right",
      })
      expect(moveTaskRightButton).toBeDisabled()
      let clickedFailed = false
      try {
        await userEvent.click(moveTaskRightButton)
      } catch (error) {
        clickedFailed = true
      }
      expect(clickedFailed).toBe(true)
    })

    test("calls the function to delete the task", async () => {
      const deleteTaskButton = screen.getByRole("button", {
        name: "delete task",
      })
      expect(deleteTaskButton).toBeInTheDocument()
      await userEvent.click(deleteTaskButton)
      expect(functionsTaskMock.onDeleteTask).toHaveBeenCalled()
      expect(functionsTaskMock.onDeleteTask).toHaveBeenCalledWith(
        defaultTask.id
      )
      expect(
        screen.queryByRole("listitem", { name: "task" })
      ).not.toBeInTheDocument()
    })
  })

  test("calls the function when clicking the checkbox", async () => {
    const checkboxElement = screen.getByRole("checkbox")
    expect(checkboxElement).toBeInTheDocument()
    await userEvent.click(checkboxElement)
    expect(functionsTaskMock.onEditTask).toHaveBeenCalled()
    expect(functionsTaskMock.onEditTask).toHaveBeenCalledWith({
      taskId: defaultTask.id,
      newTask: { ...defaultTask, completed: !defaultTask.completed },
    })
  })
})
