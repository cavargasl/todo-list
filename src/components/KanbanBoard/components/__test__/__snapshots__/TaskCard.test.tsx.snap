// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TaskCard > renders component with completed style 1`] = `
<li
  aria-label="task"
  class="flex w-full gap-2 rounded-md bg-background p-2"
>
  <input
    checked=""
    class="checkbox h-5 w-5 cursor-pointer appearance-none rounded-full border bg-black px-2 outline-none checked:border-transparent checked:bg-green-500 focus:border-inherit focus:outline-none peer"
    type="checkbox"
  />
  <p
    class="w-full peer-checked:line-through"
  >
    Task 2
  </p>
  <div
    class="flex gap-1"
  >
    <button
      aria-label="move to left"
      class="inline-flex items-center justify-center gap-2 text-nowrap font-semibold transition-all focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-foreground disabled:pointer-events-none disabled:opacity-50 text-gray-500 hover:bg-foreground hover:text-white rounded size-[20px] p-0"
    >
      <svg
        class="size-6"
        fill="none"
        stroke="currentColor"
        stroke-width="1.5"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M6.75 15.75 3 12m0 0 3.75-3.75M3 12h18"
          stroke-linecap="round"
          stroke-linejoin="round"
        />
      </svg>
    </button>
    <button
      aria-label="move to right"
      class="inline-flex items-center justify-center gap-2 text-nowrap font-semibold transition-all focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-foreground disabled:pointer-events-none disabled:opacity-50 text-gray-500 hover:bg-foreground hover:text-white rounded size-[20px] p-0"
    >
      <svg
        class="size-6"
        fill="none"
        stroke="currentColor"
        stroke-width="1.5"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M17.25 8.25 21 12m0 0-3.75 3.75M21 12H3"
          stroke-linecap="round"
          stroke-linejoin="round"
        />
      </svg>
    </button>
    <button
      aria-label="delete task"
      class="inline-flex items-center justify-center gap-2 text-nowrap font-semibold transition-all focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-foreground disabled:pointer-events-none disabled:opacity-50 text-gray-500 hover:bg-foreground hover:text-white rounded size-[20px] p-0"
    >
      <svg
        class="size-6"
        fill="none"
        stroke="currentColor"
        stroke-width="1.5"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="m14.74 9-.346 9m-4.788 0L9.26 9m9.968-3.21c.342.052.682.107 1.022.166m-1.022-.165L18.16 19.673a2.25 2.25 0 0 1-2.244 2.077H8.084a2.25 2.25 0 0 1-2.244-2.077L4.772 5.79m14.456 0a48.108 48.108 0 0 0-3.478-.397m-12 .562c.34-.059.68-.114 1.022-.165m0 0a48.11 48.11 0 0 1 3.478-.397m7.5 0v-.916c0-1.18-.91-2.164-2.09-2.201a51.964 51.964 0 0 0-3.32 0c-1.18.037-2.09 1.022-2.09 2.201v.916m7.5 0a48.667 48.667 0 0 0-7.5 0"
          stroke-linecap="round"
          stroke-linejoin="round"
        />
      </svg>
    </button>
  </div>
</li>
`;

exports[`TaskCard > renders the component  1`] = `
<li
  aria-label="task"
  class="flex w-full gap-2 rounded-md bg-background p-2"
>
  <input
    class="checkbox h-5 w-5 cursor-pointer appearance-none rounded-full border bg-black px-2 outline-none checked:border-transparent checked:bg-green-500 focus:border-inherit focus:outline-none peer"
    type="checkbox"
  />
  <p
    class="w-full peer-checked:line-through"
  >
    Task 1
  </p>
  <div
    class="flex gap-1"
  >
    <button
      aria-label="move to left"
      class="inline-flex items-center justify-center gap-2 text-nowrap font-semibold transition-all focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-foreground disabled:pointer-events-none disabled:opacity-50 text-gray-500 hover:bg-foreground hover:text-white rounded size-[20px] p-0"
    >
      <svg
        class="size-6"
        fill="none"
        stroke="currentColor"
        stroke-width="1.5"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M6.75 15.75 3 12m0 0 3.75-3.75M3 12h18"
          stroke-linecap="round"
          stroke-linejoin="round"
        />
      </svg>
    </button>
    <button
      aria-label="move to right"
      class="inline-flex items-center justify-center gap-2 text-nowrap font-semibold transition-all focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-foreground disabled:pointer-events-none disabled:opacity-50 text-gray-500 hover:bg-foreground hover:text-white rounded size-[20px] p-0"
    >
      <svg
        class="size-6"
        fill="none"
        stroke="currentColor"
        stroke-width="1.5"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M17.25 8.25 21 12m0 0-3.75 3.75M21 12H3"
          stroke-linecap="round"
          stroke-linejoin="round"
        />
      </svg>
    </button>
    <button
      aria-label="delete task"
      class="inline-flex items-center justify-center gap-2 text-nowrap font-semibold transition-all focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-foreground disabled:pointer-events-none disabled:opacity-50 text-gray-500 hover:bg-foreground hover:text-white rounded size-[20px] p-0"
    >
      <svg
        class="size-6"
        fill="none"
        stroke="currentColor"
        stroke-width="1.5"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="m14.74 9-.346 9m-4.788 0L9.26 9m9.968-3.21c.342.052.682.107 1.022.166m-1.022-.165L18.16 19.673a2.25 2.25 0 0 1-2.244 2.077H8.084a2.25 2.25 0 0 1-2.244-2.077L4.772 5.79m14.456 0a48.108 48.108 0 0 0-3.478-.397m-12 .562c.34-.059.68-.114 1.022-.165m0 0a48.11 48.11 0 0 1 3.478-.397m7.5 0v-.916c0-1.18-.91-2.164-2.09-2.201a51.964 51.964 0 0 0-3.32 0c-1.18.037-2.09 1.022-2.09 2.201v.916m7.5 0a48.667 48.667 0 0 0-7.5 0"
          stroke-linecap="round"
          stroke-linejoin="round"
        />
      </svg>
    </button>
  </div>
</li>
`;
